src/rprblender/engine/viewport_engine_2.py

```diff
@@ -0,0 +1,296 @@
+#**********************************************************************
+# Copyright 2020 Advanced Micro Devices, Inc
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#********************************************************************
+import time
+
+import pyrpr
+import bpy
+
+from rprblender.export import object
+from .viewport_engine import (
+    ViewportEngine, ViewportSettings, ShadingData, FinishRenderException
+)
+from rprblender.utils import gl
+from .context import RPRContext2
+
+from rprblender.utils import logging
+log = logging.Log(tag='viewport_engine_2')
+
+
+class ViewportEngine2(ViewportEngine):
+    _RPRContext = RPRContext2
+
+    def __init__(self, rpr_engine):
+        super().__init__(rpr_engine)
+
+        self.is_abort_render = False
+        self.rendered_image = None
+
+    def stop_render(self):
+        self.is_abort_render = True
+        super().stop_render()
+
+    def _do_render(self):
+        iteration = 0
+        time_begin = 0.0
+        update_iterations = 1
+
+        def render_update(progress):
+            # don't need to do intermediate update for 0, 1 iteration and
+            # at render finish when progress == 1.0
+            if iteration <= 1 or progress == 1.0:
+                return
+
+            if self.is_abort_render:
+                self.rpr_context.abort_render()
+                return
+
+            self._resolve()
```

Comment:
As with discussion on slack, shouldn't this NOT be happening in the callback?  Callback should set a flag like ready_for_resolve and other thread can resolve.